src/codegen/util/hash_table.cpp

```diff
@@ -0,0 +1,323 @@
+}
+
+char *HashTable::Insert(uint64_t hash) {
+  // Resize the hash table if needed
+  if (NeedsResize()) {
+    Resize();
+  }
+
+  // Acquire/allocate an entry from storage
+  Entry *entry = entry_buffer_.NextFree();
+  entry->hash = hash;
+
+  // Insert into hash table
+  uint64_t index = hash & directory_mask_;
+  entry->next = directory_[index];
+  directory_[index] = entry;
+
+  num_elems_++;
+
+  // Return data pointer for key/value storage
+  return entry->data;
+}
+
+void HashTable::BuildLazy() {
+  // Grab entry head
+  Entry *head = directory_[0];
+
+  // Clean up old directory
+  memory_.Free(directory_);
+
+  // At this point, all the lazy insertions are assumed to have completed. We
+  // can allocate a perfectly sized hash table with 50% load factor.
+  //
+  // TODO: Use sketches to estimate the real # of unique elements
+  // TODO: Perhaps change probing strategy based on estimate?
+
+  directory_size_ = NextPowerOf2(num_elems_) * 2;
+  directory_mask_ = directory_size_ - 1;
+
+  uint64_t alloc_size = sizeof(Entry *) * directory_size_;
+  directory_ = static_cast<Entry **>(memory_.Allocate(alloc_size));
+
+  PELOTON_MEMSET(directory_, 0, alloc_size);
+
+  // Now insert all elements into the directory
+  while (head != nullptr) {
+    uint64_t index = head->hash & directory_mask_;
+    Entry *next = head->next;
+    head->next = directory_[index];
+    directory_[index] = head;
+    head = next;
+  }
+}
+
+void HashTable::ReserveLazy(
+    const executor::ExecutorContext::ThreadStates &thread_states,
+    uint32_t hash_table_offset) {
+  // Determine the total number of tuples stored across each hash table
+  uint64_t total_size = 0;
+  for (uint32_t i = 0; i < thread_states.NumThreads(); i++) {
+    auto *hash_table = reinterpret_cast<HashTable *>(
+        thread_states.AccessThreadState(i) + hash_table_offset);
+    total_size += hash_table->NumElements();
+  }
+
+  // TODO: Combine sketches to estimate the true unique # of elements
+
+  // Perfectly size the hash table
+  num_elems_ = 0;
+  capacity_ = NextPowerOf2(total_size);
+
+  directory_size_ = capacity_ * 2;
+  directory_mask_ = directory_size_ - 1;
+
+  uint64_t alloc_size = sizeof(Entry *) * directory_size_;
+  directory_ = static_cast<Entry **>(memory_.Allocate(alloc_size));
+  PELOTON_MEMSET(directory_, 0, alloc_size);
+}
+
+void HashTable::MergeLazyUnfinished(HashTable &other) {
+  // Begin with the head of the linked list of entries, stored in the first
+  // directory entry
+  PELOTON_ASSERT(other.directory_[0] != nullptr);
+  auto *head = other.directory_[0];
+
+  while (head != nullptr) {
+    // Compute the index and stash the next entry in the linked list
+    uint64_t index = head->hash & directory_mask_;
+    Entry *next = head->next;
+
+    // Try to CAS in this entry into the directory
+    Entry *curr;
+    do {
+      curr = directory_[index];
+      head->next = curr;
+    } while (!::peloton::atomic_cas(directory_ + index, curr, head));
```

Comment:
I have a question about this: Do we know that `directory_[index]` is being read every time from memory and not cached in a register? Is there a rule for this?

Reply:
`atomic_cas` uses a memory fence to prevent caching,